File: Boyer-Moore/boyer_moore.cpp
#include <bits/stdc++.h> 
using namespace std;

void precomputeShifts(string patt, map <char, int> & mismatchTable) {
	int len = patt.size();
	for(int i = 0; i < len; i++) {
		char currChar = patt[i];
		int shift = max(1, len - i - 1);
		mismatchTable[currChar] = shift;
	}
}

int boyer_moore(string txt, string patt, map <char, int> & mismatchTable) {
	int skips = 0;
	for(int i = 0; i <= txt.size() - patt.size(); i += skips) {
		skips = 0;
		for(int j = patt.size() - 1; j >= 0; j--) {
			if(patt[j] != txt[i + j]) {
				if(mismatchTable[txt[i + j]] != 0) {
					skips = mismatchTable[txt[i + j]];
					break;
				}
				else {
					skips = patt.size();
					break;
				}
			}
		}
		if(skips == 0) return i;
	}
}

int main() {
	string text = "ABCABABCABAC";
    string pattern = "CAB";
    // string pattern = "BAB";
 	map <char, int> mismatchTable;
    precomputeShifts(text, mismatchTable);
    cout << "pattern found at: " << boyer_moore(text, pattern, mismatchTable);
	return 0;
}

File: Boyer-Moore/boyer_moore.java
import java.util.*;

public class boyer_moore {
	private static void precomputeShifts(String patt, HashMap <Character, Integer> mistmatchTable) {
		int len = patt.length();
		for(int i = 0; i < len; i++) {
			char currChar = patt.charAt(i);
			int shift = Math.max(1, len - 1 - i);
			mistmatchTable.put(currChar, shift);
		}
	}

	private static int doBoyerMoore(String txt, String patt, HashMap <Character, Integer> mistmatchTable) {
		int skips = 0;
		for(int i = 0; i < txt.length() - patt.length(); i += skips) {
			skips = 0;
			for(int j = patt.length() - 1; j >= 0; j--) {
				if(patt.charAt(j) != txt.charAt(i + j)) {
					if(mistmatchTable.get(txt.charAt(i + j)) != null) {
						skips = mistmatchTable.get(txt.charAt(i + j));
						break;
					} else {
						skips = patt.length();
						break;
					}
				}
			}
			if(skips == 0) return i;
		}
		return -1;
	}

	public static void main(String [] args) {
		String text = "ABCABABCABAC";
	    	String pattern = "CAB";
	    	HashMap <Character, Integer> mistmatchTable = new HashMap <> ();
	    	precomputeShifts(pattern, mistmatchTable);
    		System.out.println("pattern found at: " + doBoyerMoore(text, pattern, mistmatchTable));
	}
}
